sdks/python/apache_beam/io/iobase.py

```diff
@@ -1093,3 +1095,212 @@ def process(self, context):
+        if running_total[k] == running_total[k - 1]:
+          for j in range(k):
+            running_total[j] *= (1 - min_diff)
+      return running_total
+
+    def start_position(self):
+      return self._start
+
+    def stop_position(self):
+      return self._end
+
+    def try_claim(self, pos):
+      source_ix, source_pos = pos
+      with self._lock:
+        if source_ix > self._end[0]:
+          return False
+        elif source_ix == self._end[0] and self._end[1] is None:
+          return False
+        else:
+          self._claimed_source_ix = source_ix
+          if source_pos is None:
+            return True
+          else:
+            return self.sub_range_tracker(source_ix).try_claim(source_pos)
+
+    def try_split(self, pos):
+      source_ix, source_pos = pos
+      with self._lock:
+        if source_ix < self._claimed_source_ix:
+          # Already claimed.
+          return None
+        elif source_ix > self._end[0]:
+          # After end.
+          return None
+        elif source_ix == self._end[0] and self._end[1] is None:
+          # At/after end.
+          return None
+        else:
+          if source_ix > self._claimed_source_ix:
+            # Prefer to split on even boundary.
+            split_pos = None
+            ratio = self._cumulative_weights[source_ix]
+          else:
+            # Split the current subsource.
+            split = self.sub_range_tracker(source_ix).try_split(
+                source_pos)
+            if not split:
+              return None
+            split_pos, frac = split
+            ratio = self.local_to_global(source_ix, frac)
+
+          self._end = source_ix, split_pos
+          self._cumulative_weights = [min(w / ratio, 1)
+                                      for w in self._cumulative_weights]
+          return (source_ix, split_pos), ratio
+
+    def set_current_position(self, pos):
+      raise NotImplementedError('Should only be called on sub-trackers')
+
+    def position_at_fraction(self, fraction):
+      source_ix, source_frac = self.global_to_local(fraction)
+      if source_ix == len(self._sources):
+        return (source_ix, None)
+      else:
+        return (source_ix,
+                self.sub_range_tracker(source_ix).position_at_fraction(
+                    source_frac))
+
+    def fraction_consumed(self):
+      with self._lock:
+        return self.local_to_global(
+            self._claimed_source_ix,
+            self.sub_range_tracker(self._claimed_source_ix)
+            .fraction_consumed())
+
+    def local_to_global(self, source_ix, source_frac):
+      cw = self._cumulative_weights
+      return cw[source_ix] + source_frac * (cw[source_ix + 1] - cw[source_ix])
+
+    def global_to_local(self, frac):
+      if frac == 1:
+        return (len(self._sources), 0)
+      else:
+        cw = self._cumulative_weights
+        source_ix = bisect.bisect(cw, frac) - 1
+        return (source_ix,
```

Comment:
Please add comments to clarify here and other places that involve non-obvious arithmetic.